### PR TITLE
[Merged by Bors] -  feat: `iInter_sigma`, `biInter_sigma`, `biSup_sigma`, `biInf_sigma`

### PR DESCRIPTION
**Pull request by llllvvuu**
similar to `prod_sigma`, `prod_sigma'`, `iInf_sigma`, `iInf_sigma'`, `iSup_sigma`, `iSup_sigma'`

From PFR

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

File: Mathlib/Data/Finset/Sigma.lean [hunk 1]
```diff
@@ -109,6 +109,9 @@ theorem inf_sigma [SemilatticeInf β] [OrderTop β] :
   @sup_sigma _ _ βᵒᵈ _ _ _ _ _
 #align finset.inf_sigma Finset.inf_sigma
 
+theorem _root_.Set.biInter_finsetSigma (s : Finset ι) (t : ∀ i, Finset (α i))
+    (u : ∀ i, α i → Set β) : ⋂ ij ∈ s.sigma t, u ij.1 ij.2 = ⋂ i ∈ s, ⋂ j : t i, u i j := by aesop
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Sigma.lean:113]: This statement looks odd to me; shouldn't it be the following for symmetry?
```suggestion
    (u : ∀ i, α i → Set β) : ⋂ ij ∈ s.sigma t, u ij.1 ij.2 = ⋂ i ∈ s, ⋂ j ∈ t i, u i j := by aesop
```

File: Mathlib/Data/Set/Sigma.lean [hunk 2]
```diff
@@ -132,6 +133,9 @@ theorem sigma_inter_sigma : s₁.sigma t₁ ∩ s₂.sigma t₂ = (s₁ ∩ s₂
   simp [and_assoc, and_left_comm]
 #align set.sigma_inter_sigma Set.sigma_inter_sigma
 
+theorem biInter_sigma (s : Set ι) (t : ∀ i, Set (α i)) (u : ∀ i, α i → Set ι') :
+    ⋂ ij ∈ s.sigma t, u ij.1 ij.2 = ⋂ i ∈ s, ⋂ j : t i, u i j := by aesop
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Sigma.lean:137]: Same here, I hadn't noticed
```suggestion
    ⋂ ij ∈ s.sigma t, u ij.1 ij.2 = ⋂ i ∈ s, ⋂ j ∈ t i, u i j := by aesop
```

**Review by eric-wieser — CHANGES_REQUESTED** [review]
These lemmas feel rather contrived; I think we're missing something more general, like:
```lean
theorem _root_.Set.iInter_sigma (s : (Σ i, α i) → Set β) :
  ⋂ ia, s ia = ⋂ i, ⋂ a, s ⟨i, a⟩ := by ext; simp
```
and similarly for `prod` and `pi`.

Whatever we decide, we certainly shouldn't add the `Inter` versions without also adding matching `Union` versions.

**Comment by YaelDillies** [comment]
I think your suggested lemmas are good, but can't replace the ones here. Eg your lemmas would turn
* `⋂ ij ∈ s.sigma t, f ij` into
* `⋂ i j (_ : ⟨i, j⟩ ∈ s.sigma t), f ⟨i, j⟩`, which will be further simplified (due to `Set.mem_sigma_iff`) to
* `⋂ i j (_ : i ∈ s ∧ j ∈ t), f ⟨i, j⟩`. `Set.iInter_and` could then turn that into
* `⋂ i j (_ : i ∈ s) (_ : j ∈ t), f ⟨i, j⟩` but note that it's not simp.

And now you're in a pickle because swapping the second and third `⋂` is the most annoying common swap I've encountered.

**Comment by eric-wieser** [comment]
> Also note that we want iInf and iSup lemmas too.

Agreed!

File: Mathlib/Data/Set/Lattice.lean [hunk 3]
```diff
@@ -903,6 +903,20 @@ theorem iInter_comm (s : ι → ι' → Set α) : ⋂ (i) (i'), s i i' = ⋂ (i'
   iInf_comm
 #align set.Inter_comm Set.iInter_comm
 
+theorem iUnion_sigma {γ : α → Type*} (s : Sigma γ → Set β) : ⋃ ia, s ia = ⋃ i, ⋃ a, s ⟨i, a⟩ :=
+  iSup_sigma
+
+theorem iUnion_sigma' {γ : α → Type*} (s : ∀ i : α, γ i → Set β) :
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Lattice.lean:909]:
```suggestion
theorem iUnion_sigma' {γ : α → Type*} (s : ∀ i, γ i → Set β) :
```

**Comment by llllvvuu** [comment]
I created both the bounded and unbounded variants (except for in the case of `iInf_sigma` which already exists so I only created `biInf_sigma`).

Indeed I generally experienced some friction using the unbounded variants in bounded expressions.

I also kept both the `Set` and `Finset` variants since I experienced some friction using `Set` lemmas in `Finset` expressions (needed a few explicit coercions AFAICT). But instead of keeping the `Finset.univ` variants I made #8995.

File: Mathlib/Data/Set/Sigma.lean [hunk 4]
```diff
@@ -132,6 +133,38 @@ theorem sigma_inter_sigma : s₁.sigma t₁ ∩ s₂.sigma t₂ = (s₁ ∩ s₂
   simp [and_assoc, and_left_comm]
 #align set.sigma_inter_sigma Set.sigma_inter_sigma
 
+theorem _root_.biSup_sigma [CompleteLattice ι'] (s : Set ι) (t : ∀ i, Set (α i))
+    (f : Sigma α → ι') : ⨆ ij ∈ s.sigma t, f ij = ⨆ (i ∈ s) (j ∈ t i), f ⟨i, j⟩ :=
+  by apply eq_of_forall_ge_iff; intro; constructor <;> simp_all
```
Comment (llllvvuu) [inline Mathlib/Data/Set/Sigma.lean:138]: `aesop` was messing with this theorem statement (i think due to `unhygienic` pulling in implicit variables), so i expanded them all down to `simp`/`simp_all`

File: Mathlib/Data/Finset/Sigma.lean [hunk 5]
```diff
@@ -109,6 +109,38 @@ theorem inf_sigma [SemilatticeInf β] [OrderTop β] :
...
+    (f : Sigma α → β) : ⨆ ij ∈ s.sigma t, f ij = ⨆ (i ∈ s) (j ∈ t i), f ⟨i, j⟩ :=
+  by simp_rw [← Finset.iSup_coe, Finset.coe_sigma, biSup_sigma]
+
+theorem _root_.biSup_finsetSigma' [CompleteLattice β] (s : Finset ι) (t : ∀ i, Finset (α i))
+    (f : ∀ i, α i → β) : ⨆ (i ∈ s) (j ∈ t i), f i j = ⨆ ij ∈ s.sigma t, f ij.fst ij.snd :=
+  Eq.symm $ biSup_finsetSigma _ _ _
+
+theorem _root_.biInf_finsetSigma [CompleteLattice β] (s : Finset ι) (t : ∀ i, Finset (α i))
+    (f : Sigma α → β) : ⨅ ij ∈ s.sigma t, f ij = ⨅ (i ∈ s) (j ∈ t i), f ⟨i, j⟩ :=
+  @biSup_finsetSigma _ _ βᵒᵈ _ _ _ _
```
Comment (YaelDillies) [inline Mathlib/Data/Finset/Sigma.lean:122]:
```suggestion
  biSup_finsetSigma (β := βᵒᵈ) _ _ _
```

File: Mathlib/Data/Finset/Sigma.lean [hunk 6]
```diff
@@ -109,6 +109,38 @@ theorem inf_sigma [SemilatticeInf β] [OrderTop β] :
...
+    (u : ∀ i, α i → Set β) : ⋃ i ∈ s, ⋃ j ∈ t i, u i j = ⋃ ij ∈ s.sigma t, u ij.fst ij.snd :=
+  biSup_finsetSigma' _ _ _
+
+theorem _root_.Set.biInter_finsetSigma (s : Finset ι) (t : ∀ i, Finset (α i))
+    (u : Sigma α → Set β) : ⋂ ij ∈ s.sigma t, u ij = ⋂ i ∈ s, ⋂ j ∈ t i, u ⟨i, j⟩ :=
+  biInf_finsetSigma _ _ _
+
+theorem _root_.Set.biInter_finsetSigma' (s : Finset ι) (t : ∀ i, Finset (α i))
+    (u : ∀ i, α i → Set β) : ⋂ i ∈ s, ⋂ j ∈ t i, u i j = ⋂ ij ∈ s.sigma t, u ij.1 ij.2 :=
+  biInf_finsetSigma' _ _ _
```
Comment (YaelDillies) [inline Mathlib/Data/Finset/Sigma.lean:142]: Sorry, I should have told you to keep `f`
```suggestion
theorem _root_.Set.biUnion_finsetSigma (s : Finset ι) (t : ∀ i, Finset (α i))
    (f : Sigma α → Set β) : ⋃ ij ∈ s.sigma t, f ij = ⋃ i ∈ s, ⋃ j ∈ t i, f ⟨i, j⟩ :=
  biSup_finsetSigma _ _ _

theorem _root_.Set.biUnion_finsetSigma' (s : Finset ι) (t : ∀ i, Finset (α i))
    (f : ∀ i, α i → Set β) : ⋃ i ∈ s, ⋃ j ∈ t i, f i j = ⋃ ij ∈ s.sigma t, f ij.fst ij.snd :=
  biSup_finsetSigma' _ _ _

theorem _root_.Set.biInter_finsetSigma (s : Finset ι) (t : ∀ i, Finset (α i))
    (f : Sigma α → Set β) : ⋂ ij ∈ s.sigma t, f ij = ⋂ i ∈ s, ⋂ j ∈ t i, f ⟨i, j⟩ :=
  biInf_finsetSigma _ _ _

theorem _root_.Set.biInter_finsetSigma' (s : Finset ι) (t : ∀ i, Finset (α i))
    (f : ∀ i, α i → Set β) : ⋂ i ∈ s, ⋂ j ∈ t i, f i j = ⋂ ij ∈ s.sigma t, f ij.1 ij.2 :=
  biInf_finsetSigma' _ _ _
```

File: Mathlib/Data/Set/Sigma.lean [hunk 7]
```diff
@@ -132,6 +133,38 @@ theorem sigma_inter_sigma : s₁.sigma t₁ ∩ s₂.sigma t₂ = (s₁ ∩ s₂
   simp [and_assoc, and_left_comm]
 #align set.sigma_inter_sigma Set.sigma_inter_sigma
 
+theorem _root_.biSup_sigma [CompleteLattice ι'] (s : Set ι) (t : ∀ i, Set (α i))
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Sigma.lean:136]: `ι'` is not a great name for a lattice! Can you use `β` instead?
```suggestion
theorem _root_.biSup_sigma [CompleteLattice ι'] (s : Set ι) (t : ∀ i, Set (α i))
```

**Comment by YaelDillies** [comment]
I must say it would be much easier for us to see what's going on if you did not force-push every single commit.

**Comment by llllvvuu** [comment]
> I must say it would be much easier for us to see what's going on if you did not force-push every single commit.

Sorry about that; I changed it back to a full history and will be appending only going forward.

File: Mathlib/Data/Set/Sigma.lean (hunk 4 above)
Comment (ericrbg) [reply Mathlib/Data/Set/Sigma.lean:140]: what do you mean by this? the theorem statement should be fully evaluated before entering tactic mode

**Review by ericrbg — COMMENTED** [review]
Is there a more systematic way to turn these sort of statements into general statements about `And` & `biSup`, or such like? (i.e. going through `Set.mem_sigma_iff` instead).

File: Mathlib/Data/Set/Sigma.lean (hunk 4 above)
Comment (llllvvuu) [reply Mathlib/Data/Set/Sigma.lean:140]: Screenshot 2024-02-09 at 1 53 53 PM
Screenshot 2024-02-09 at 1 54 15 PM

`aesop` introduces `x` as an implicit argument, which is undesired. I don't fully understand the mechanics behind this, but `clear x; aesop` works

**Comment by llllvvuu** [comment]
@ericrbg could you describe what you're envisioning a bit more? The following proof is one based on `mem_sigma_iff` and `iSup_and`:

```lean
theorem _root_.biSup_sigma (s : Set ι) (t : ∀ i, Set (α i)) (f : Sigma α → β) :
    ⨆ ij ∈ s.sigma t, f ij = ⨆ (i ∈ s) (j ∈ t i), f ⟨i, j⟩ := by
  simp? [iSup_sigma, iSup_and] says simp only [mem_sigma_iff, iSup_and, iSup_sigma]
  conv => { rhs; congr; ext; rw [iSup_comm] }
```

By "systematic" / "general", should I make that `conv` step into a lemma?

File: Mathlib/Data/Set/Sigma.lean (hunk 4 above)
Comment (ericrbg) [reply Mathlib/Data/Set/Sigma.lean]: Oh wow, that's unhelpful, yes. Maybe worth raising this as an aesop issue

**Comment by ericrbg** [comment]
Maybe the best way is to make a lemma for the set of sigmas, and then possibly specialise that to this case?

**Comment by PatrickMassot** [comment]
@YaelDillies, @ericrbg: What is the status of this PR?

**Comment by ericrbg** [comment]
I'd like to see these more general lemmas about sets of Sigmas first

**Comment by llllvvuu** [comment]
What is the "set of sigmas"? I tried `Set (Sigma a)` but it doesn't yield anything useful:
```lean
theorem _root_.biSup_sigma (s : Set (Sigma α)) (f : Sigma α → β) :
    ⨆ ij ∈ s, f ij = ⨆ (i) (j) (_ : ⟨i, j⟩ ∈ s), f ⟨i, j⟩ :=
  iSup_sigma
```
In order to commute `(j)` to the right, I have to be able to break up the `(_ : ⟨i, j⟩ ∈ s)` for which I need `s` to be some `S.sigma T` as the "un-general" version stated.

Could I get an example of a possible theorem statement?

**Comment by ericrbg** [comment]
```lean
nonrec theorem biSup_sigma {γ β} {α : γ → Type*} [CompleteLattice β] (s : Set (Sigma α)) (f : Sigma α → β) :
  ⨆ ij ∈ s, f ij = ⨆ (i : γ) (_ : ∃ t, ⟨i, t⟩ ∈ s) (j : α i) (_ : ⟨i, j⟩ ∈ s), f ⟨i, j⟩ :=
  eq_of_forall_ge_iff fun _ ↦ ⟨by simp_all, by aesop⟩


variable {β : Type*} [CompleteLattice β] {ι : Type*} {α : ι → Type*}

nonrec theorem _root_s.biSup_sigma (s : Set ι) (t : ∀ i, Set (α i)) (f : Sigma α → β) :
    ⨆ ij ∈ s.sigma t, f ij = ⨆ (i ∈ s) (j ∈ t i), f ⟨i, j⟩ := by
  convert biSup_sigma (s.sigma t) _
  aesop
  rename_i x
  -- this seems like a nonsense goal
```

maybe something like this? apologies these are not very tidy, and honestly the goal doesn't inspire confidence at the end, but something of that form may be okay. if you don't think this works and you don't have any other similar ideas, I'm happy to maintainer merge this tomorrow

**Comment by llllvvuu** [comment]
I think `(_ : ∃ t, ⟨i, t⟩ ∈ s)` and `(i ∈ s)` are probably not equivalent since `Nonempty` was not specified

**Comment by YaelDillies** [comment]
> @YaelDillies, @ericrbg: What is the status of this PR?

I'm happy with it and I don't understand what theorems eric is after.

**Comment by ericrbg** [comment]
maintainer merge

**Pull request closed without merging**